Header Only Library

cfis opened the pull request:
As discussed, this commit updates Rice to be a header only library. It also makes use of recent C++17 features to simplify and update the library. Each commit contains a description of its changes.  

The idea is users include/rice/rice.hpp via #include <rice/rice.hpp>. That version of rice.hpp include all code combined together. It is  generated by a ruby script (run rake rice_hpp).  There is also a rice.hpp file in the rice directory. That one is a valid C++ header file that is just a list of includes. I use that one for development. It is also the file the rake task processes to generate the single combined file.

I have updated the build system to just use mkmf, although I did have to hack it a bit to get it to compile unittest to an executable. To use:

rake test
rake build (to just compile unittest and samples)
rake clean (to clean up unitest and samples)
rake rice_hpp to make the single header

Note that mkmf-rice.rb is still in use - its job is still to link to the standard C++ library and the find rice.hpp.

I have tested on Windows with MSVC and Mingw, on MacOS with the latest Xcode and on Fedora with g++.

I also did some small documentation updates, but more work is needed on that.

Last, these change are not fully backwards compatible due to changes in the the way To_Ruby and From_Ruby are implemented. Its easy to change existing code, but that would have to be documented. It might also be possible to have a backwards_compat.hpp header file. I also would like to change how Arg are defined before doing a release, but that's for a separate discussion.

jasonroelofs commented:
Awesome, thanks for all this work! Couple of initial comments / questions:

* First, can I get collaborator access to your repo? I'd like to push some changes to help get CI passing.
* I see that we have `rice/rice.hpp` as well as `include/rice/rice.hpp`, are both necessary or is one an experiment
* Related, I would like to see `rice/rice.hpp` be what we build, but would like to know your thinking behind a file with includes vs concatenating everything together.

Will start working through this and comment as I come across things.

jasonroelofs commented on README.md:
```diff
@@ -278,20 +271,29 @@ Ruby types:
-    // ...
-  }
+    template<>
+    structure From_Ruby<Foo>
```
Did `structure` get added as a keyword or is this supposed to be `struct` that got autocompleted?

cfis replied:
Good catch - should be struct. Will fix that.

jasonroelofs commented on rice.gemspec:
```diff
@@ -39,49 +35,19 @@ of the Ruby C API.
 
-    # C++ source files
+    # rice.hpp
+    'include/rice.hpp',
```
This probably should be `include/rice/rice.hpp`.

cfis replied:
Yup. I had started with #include <rice.hpp> and then decided #include <rice/rice.hpp> was a better idea for any future changes additions (like adding stl support or a backward compatibility header).

jasonroelofs commented on rice.gemspec:
```diff
@@ -100,20 +66,16 @@ of the Ruby C API.
-  s.files = patterns.collect { |p| Dir.glob(p) }.flatten
-
-  s.required_ruby_version = ">= 2.4"
+  s.required_ruby_version = ">= 2.7"
```
Can we bump this to `2.5`? Trying to support all versions that the Ruby team supports.

cfis replied:
Sure. Note that Ruby 2.5 will no longer be supported as on March 31 though.

jasonroelofs commented on rice/Arg.ipp:
```diff
@@ -59,69 +55,37 @@ namespace Rice {
 
       //! Our saved default value
-      type_base* defaultValue;
+      std::any defaultValue_;
```
Love the C++ has this type built in now!

cfis replied:
Yes, very handy!

jasonroelofs commented:
Sorry if this is explained in the commits somewhere, but didn't see an obvious commit for it, but could you explain why `to_ruby` and `from_ruby` needed to change structures?

Also, and this may not be internally consistent, the `details` namespace is meant for things internal to Rice, can we move `To_Ruby` and `From_Ruby` out of that namespace?

jasonroelofs commented:
Oh, and I see now how `rice/rice.hpp` and `include/rice/rice.hpp` interact and I guess what's the benefit or reason for using a fully concatenated single file vs `rice/rice.hpp`?

cfis commented:
Some comments inline above. As for your other questions:

* I added you as a collaborator
* On rice/rice.hpp versus include/rice/rice.hpp. For development purposes its easier to use rice/rice.hpp because you can change code, recompile and test. If instead you use include/rice/rice.hpp you change code but then need to rebuild the header before compiling/testing. But for users of the library include/rice/rice.hpp is more convenient. In fact you could copy it into your project and not need to install the rice gem at all if you wanted. For example, that's how I use https://github.com/nlohmann.
*  On the From_Ruby/To_Ruby, a few thoughts. First I was really trying to layer the code so that code in the Rice namespace knows about detail, but detail doesn't know about Rice. That's not 100% possible, but it helped remove some circular dependencies which enabled having a single header file. Since the code in detail needs to know about From_Ruby/To_Ruby it should to be defined in detail. Second, for what its worth, pybind11 does the same thing with conversion code in the detail namespace with the note that "yes we know we are asking you to override something in detail namespace". Third, I changed the code to use template structures because that code requires partial specialization support (not allowed with template functions). Note in the original code to_ruby and from_ruby under the hood, in some cases, would also map to structures themselves (To_Ruby_/From_Ruby_) but the user api was to overload fully specialize those functions. I thought it was more consistent to just do it one way. Anyway, we could add the same mapping back in to mimic the old api if you want.

cfis commented:
One thing to point out. The old code used CRTP so that if you called Module#define_method you would get back a module and if you called Class#define_method you would get back a Class. That's obviously very important, so the hacky replacement was to define the methods in Module (per usual) and then in Class and Data_Type, include this header file:

#include "shared_methods.hpp"

If you look at that file, it defines "define_method" to call Module#define_method. And then it returns *this but with a return type of auto&. Thus when that file is included in Class, the method returns a Class. When that file is included in Data_Type it returns a Data_Type. You can almost get away with not including it at all in Data_Type except if you want to call define_iterator. That used to be defined on Class, but I couldn't come up with any use case for defining an iterator on a pure Ruby class (can you?).  Since that seemed to be something you would only define on a C++ class you are wrapping I moved that method to Data_Type. But that means that if you do Data_Type<Foo>.define_method(...).define_iterator(...) then define_method needs to return a Data_Type.

jasonroelofs commented:
As I work with this I think the concatenated file isn't going to work for two reasons. First, compilation errors and test failures get harder to track down, cause you get a line number in the concatenated file and then need to find it in the real source. Second, even with some scripts it's too easy to push changes that don't re-generate the file.

That isn't to say the idea of a concatenated file is itself a bad one, but I don't think it works for developing the library itself. Instead, for the case you mentioned of being able to drop the thing verbatim into another code base, we could document how to do this, or we could even expose it as a simple CLI.

Thoughts?

Also, great change with Arg, didn't even think to apply the pattern there when I first approached variardic templates and parameters!

cfis commented:
Yes, I agree. The concatenated file is not good for development. But for external use, I think it would pretty handy. If it was checked into github its easy to just grab and download. The alternative would be, as you mention, to install the gem and then run a rake task or command line to create it.

Currently the tests use the concatenated file to verify that it works. I think we should keep this because if the concatenated file works then the includes file works, but the reverse is not always true.

cfis commented:
Ok, I now remember why I specified Ruby 2.7 as the minimum version. Its because experimental C++ support was added for 2.7 starting with this commit (https://github.com/ruby/ruby/commit/c3f03da4944241f46233ce159068f35c5bd5f9b2) and then some after that. If we want to support 2.5 or 2.6 then we would need to just cut and paste that new code into Rice.

cfis commented:
Ok tests now pass on all platforms...

jasonroelofs commented:
I see you're still making some significant changes so I won't plan on merging this PR until you are comfortable with it.

Things to still figure out (can I can put some thought into this): A better way to build, test, and handle the combined file, a possible mapping of the old to_/from_ruby to the new format.

Other than that, this is great work, thanks so much for taking the time to pull this all together.

jasonroelofs commented:
I have some concerns about the "keepAlive", mainly that I don't think `Arg` is the right place for this to live. I would expect any container to be responsible for ensuring that any values it owns get marked. I thought something like this already existed in Rice, but I'm not seeing anything now that I look.

What are your thoughts on something more like:

```
  define_class<ListenerContainer>("ListenerContainer")
    .define_method("add_listener", &ListenerContainer::addListener)
    .define_mark_func(&ListenerContainer::gcMark)
```

cfis commented:
Yeah that was my first thought too. But as far as I can see, it doesn't work. Take a look at the test_keep_alive.cpp example - its code that used to live in test_data_type I just separated it out to a new file.

The example has a C++ class called ListenerContainer (well it used to be called Handler so I did rename it), which contains Listener objects. If you define a mark function for the ListenerContainer the mark function will get called with ```ListenerContainer* container```. But the container has no knowledge of the Ruby Listener objects that are wrapping the C++ Listener objects that it contains. So there is nothing it can mark.

The other way around doesn't work either, you can't add a mark function to Listener because it never gets called with the way the code is currently written. The Ruby GC just immediately cleans it up - meaning free is called without a call to mark.

Thus the current solution. In the same way you can tell Rice when ownership of an object is transferred (Rice used to say Ruby always owns a wrapped object but its easy to come up with examples where that is a bad assumption), you can now tell it when it has to tie the lifetime of one Ruby object to another. The storage of this is done via a std::vector<VALUE> on the Wrapper class which is what Ruby is now storing in a TypedData object.

Note that PyBind also has this issue - which is where I found inspiration. Its solution is a bit more complicated, it adds a separate keep_alive option to the end of the argument list which lets you tie the lifetime of one argument to another or to the receiving object. That seemed a bit of overkill to me, so I simplified to just allowing tying an argument to the receiving object (ie., the container use case).

Open to other ideas if you have any!

cfis commented:
Quick status update. I think this is about done, but I'm running into a crash with GCC. GCC is getting confused by the Wrapper object hierarchy.  When given a pointer to Wrapper<T> which is really a WrapperPointer<T>, when it calls get() its calls the pure virtual method on Wrapper<T> instead of the overridden one on WrapperPointer<T>. Note this only happens when Director objects are being stored, not with other types of objects. Somehow when we ask for a Wrapper<Listener> but what is really stored is Wrapper<ListenerDirector> the pure virtual method gets called. 

I almost never say this, but this looks like a compiler bug to me since MSVC is fine and GCC is fine with -Og, -O1 or no O at all. Anyway, to fix I think requires removing the template from the base class and return void pointers instead. Boo.

jasonroelofs commented:
Keep alive: Ah, yes, I guess I hadn't thought too deeply about how such a thing (`define_mark_func`) would have to work. Thanks for the detailed reasoning, this does sound like the best solution, and nothing wrong following pybind's solution!

GCC crash, that is weird. Don't really have any other ideas there sorry.

Do see some odd crashes on CI, can't seem to reproduce that locally but the stacktraces seem to implicate the embed_ruby somehow. Thoughts on that?

cfis commented:
Embed ruby yes. Its a Ruby 3 bug which I reported:

https://bugs.ruby-lang.org/issues/17643

Need to remove GC.start calls. So for the moment I'm just working on Ruby 2.7 and will resolve this issue once that is clean.

cfis commented:
Jason - I think this is ready to be merged, but not released. Maybe to a dev branch? Before releasing, I would like to deal with #147, #148 (which I suspect will require some changes to the library) and also get a bit of real-world work against the library I'm currently wrapping with PyBind11.

jasonroelofs commented:
Cool. Sounds good. I forget if you mentioned this already, but the one test failure, is that known / expected?

My plan right now is to create a 3.x branch so that we can keep that version moving with anything necessary, and pull this into master. Then update the docs to be really clear that there are now two paths to the library and make sure the backwards incompatible changes are well called out. I might also take this time to revisit the documentation as we'll want to keep a 3.x and 4.x of those available as well.

jasonroelofs commented:
Ha, and I guess I was looking at old data. As soon as I posted my comment, the view updated and is now all green. Never mind that!

ankane commented:
Hey @cfis, just tried on two smaller projects. I'm not very good at C++, so some of these may be user-error.

- Midas - [Diff](https://github.com/ankane/midas/compare/rice-header-only) - [CI](https://github.com/ankane/midas/actions/runs/612127165)
- Faiss - [Diff](https://github.com/ankane/faiss/compare/rice-header-only) - [CI](https://github.com/ankane/faiss/actions/runs/612160008)

Great
- No compile time for Rice :tada:
- Single include (`rice/rice.hpp`) instead of many
- Think this will make it easy to cross-compile gems that use Rice (to ship precompile gems - still need to test)

Issues
- Seeing `error: variable has incomplete type 'Rice::Class'` on Mac (see CI for full output)
- Seeing redefinition errors when `rice.hpp` is included more than once - think it's missing `#pragma once`
- Seeing `terminate called after throwing an instance of 'std::bad_any_cast'  - what(): bad any_cast` when running tests for Faiss

Other thoughts
- Does the `Arg` change leave room to add support for keyword arguments in the future?
- Re #99, is it better to use lambda instead pointer to lambda for methods going forward?

For the next round, I can test on larger projects like OR-Tools and Torch.rb.

ankane commented:
fwiw, locally with Faiss, not seeing `bad any_cast`, but instead seeing:

`wrong number of arguments` errors for instance methods
- caller `_add(objects.shape[0], objects)`
- definition `.define_method("_add", *[](faiss::Index &self, int64_t n, Rice::Object o) {`

`wrong argument type Faiss::IndexFlatL2 (expected Faiss::Index)` when `Faiss::IndexFlatL2` is an instance of `Faiss::Index`
  - definition `Rice::define_class_under<faiss::IndexFlatL2, faiss::Index>(m, "IndexFlatL2")`

cfis commented:
@jasonroelofs - I worked around the Ruby 3.0 issue by not enabling GC stress for tests. Having said that, the addition of the define_attr has broken the build on gcc , so I need to fix that (thought I had with the enum fix but apparently not).

@ankane - Thanks for those CI build links, will take a look at those. As far as the warning, yeah, that's wrong but I'm not sure how to fix (there is a todo in the notes). This has to do with implicit_casts. Previously, if an implict cast happened it caused a memory leak (create a new object that is never freed). That in fact is still an issue with a pointers, and with references you get a reference to a local object so would get a crash. I need to figure out something on that.

cfis commented:
Ok back mostly to green. The failure you see is on Ruby 2.7 when GC.stress is on. There is a pre-existing Array test that stores Ruby objects in a vector, so std::vector<Object>.

The problem with that is the Objects sometimes get garbage collected during the test  and thus the failure. I'm not sure its fixable. Obviously std::vector doesn't have a mark function, thus you can't call rb_register_address. Nor could you call rb_register_object. Nor do I know of a way to tell Ruby "please call this standalone mark function" so we could loop over the vector in each sweep phase." 

Anyway, I'll take a look at @ankane failures and see what's up with them.

jasonroelofs commented:
I'm not sure what's going on with the MacOS 2.7 test case. I can't reproduce the failure locally, but it's somehow getting that iterator out of order?

Anyway, I've set up a new `3.x` branch and it's probably time to just merge this work into `master` and continue iterating on things there, if that's ok with you @cfis? Also would you like to have direct committer access to master or make further changes as further PRs?

cfis commented:
Hey Jason - Its a GC issue. You also sometimes see a crash on Ubuntu with that code. The problem is this:

std::vector<Object> v;

When GC.stress is true, then in theory each time a Ruby object is created the GC should run. The objects in that vector are not protected by a mark function (and I don't see anyway they can be - I don't even see how to protect Object itself - it had a mark function but nothing actually called it) - thus if you are unlucky you get invalid data or lucky a crash (I'd rather have the crash). If you have any ideas how to fix - would love to hear them. Otherwise, we should probably remove that test. But someone undoubtedly already does this, or is going to do this, and as a result will get random crashes when their program is under load.

Merging to master sounds good - and I'd prefer commit access. Thanks!

jasonroelofs commented:
The GC description makes sense, though in this case it's pretty consistently failing on just MacOS and just Ruby 2.7, so it's kind of a weird thing. It's also kind of a weird test that might need to be revisited anyway.

Alright, lets get this merged, and I'll get you access to the repo directly! Thanks again for all your work here!

jasonroelofs commented:
Oh, actually there is a new error that we should fix here first before merging.

```
map.cpp:54:61: error: call of overloaded ‘protect(VALUE (&)(VALUE, VALUE), const Rice::Object&, const Rice::Object&)’ is ambiguous
   54 |         return protect(rb_assoc_new, pair.first, pair.second);
```

cfis commented:
That's fixed. Notice the failing test failed on 3 platforms this time, so it is bit random.

jasonroelofs commented:
Cool, and the fact that some of these failures are outright segfaulting on that same test do lead to it being GC. Ok, time to merge this and we'll look to see what we can do about getting this test to be a little more reliable.
